presto-main/src/main/java/io/prestosql/execution/resourcegroups/InternalResourceGroup.java

```diff
@@ -858,7 +893,10 @@ private boolean canRunMore()
     {
         checkState(Thread.holdsLock(root), "Must hold lock");
         synchronized (root) {
-            if (cpuUsageMillis >= hardCpuLimitMillis) {
+            long cpuUsageMillis = cachedResourceUsage.getCpuUsageMillis();
+            long memoryUsageBytes = cachedResourceUsage.getMemoryUsageBytes();
+
+            if ((cpuUsageMillis >= hardCpuLimitMillis) || (memoryUsageBytes > softMemoryLimitBytes)) {
```

Comment:
This gets me thinking about the soft vs hard concept again. I believe the way we're using CPU is here is also like a 'soft' limit, since at the hard limit we will kill the query.

Reply:
A running query is killed for its cpu usage _only_ when query's cpu crosses `query.max-cpu-time`. I don't think we ever kill a running query based on the resource group usage right now. Current behavior is that overshooting `hardCpuLimitMillis` prevents all incoming queries in that group from running. Overshooting `softCpuLimitMillis` practically lowers the concurrency limit. Do you think we should change this behavior? If we do, I think that such a change will be independent to the work in this PR. LMK if I missed anything.

Reply:
Ah no, you're right. I misremembered/misunderstood.